demos/adjoint/adjoint_demo.py

```diff
@@ -0,0 +1,342 @@
+# $$ \text{Objective Functional}= \int_{\Omega}(T - T_{\text{obs}}) ^ 2 \, dx \\
+#                  +\alpha_u\, \frac{D_{T_{obs}}}{N\times D_{u_{obs}}}\sum_{i}\int_{\partial \Omega_{\text{top}}}(u - u_{\text{obs}}) \cdot(u - u_{\text{obs}}) \, ds \\
+#                  +\alpha_s\, \frac{D_{T_{obs}}}{D_{\text{smoothing}}}\int_{\Omega} \nabla(T_{ic} - T_{\text{average}}) \cdot \nabla(T_{ic} - T_{\text{average}}) \, dx \\
+#                  +\alpha_d\, \frac{D_{T_{obs}}}{D_{\text{damping}}}\int_{\Omega}(T_{ic} - T_{\text{average}}) ^ 2 \, dx $$
+
+# With the three *normlisation terms* of:
+# + $D_{\text{damping}} = \int_{\Omega} T_{\text{average}}^2 \, dx$,
+# + $D_{\text{smoothing}} = \int_{\Omega} \nabla T_{\text{obs}} \cdot \nabla T_{\text{obs}} \, dx$,
+# + $D_{T_{obs}} = \int_{\Omega} T_{\text{obs}} ^ 2 \, dx$, and
+# + $D_{\text{damping}} = \int_{\partial \Omega_{\text{top}}} u_{\text{obs}} \cdot u_{\text{obs}} \, ds$
+#
+# which we define with the following definitions of `objective`
+
+# +
+# Define the component terms of the overall objective functional
+damping = assemble((Tic - Taverage) ** 2 * dx)
+norm_damping = assemble(Taverage**2 * dx)
+smoothing = assemble(dot(grad(Tic - Taverage), grad(Tic - Taverage)) * dx)
+norm_smoothing = assemble(dot(grad(Tobs), grad(Tobs)) * dx)
+norm_obs = assemble(Tobs**2 * dx)
+norm_u_surface = assemble(dot(uobs, uobs) * ds_t)
+
+# Temperature misfit between solution and observation
+t_misfit = assemble((T - Tobs) ** 2 * dx)
+
+# Weighting terms
+alpha_u = 1e-1
+alpha_d = 1e-2
+alpha_s = 1e-1
+
+# Define the overall objective functional
+objective = (
+    t_misfit +
+    alpha_u * (norm_obs * u_misfit / timesteps / norm_u_surface) +
+    alpha_d * (norm_obs * damping / norm_damping) +
+    alpha_s * (norm_obs * smoothing / norm_smoothing)
+)
+# -
+
+# Defining the Reduced Functional and Optimisation Method
+# ===============================
+#
+# In optimisation terminology, a reduced functional is a functional that takes a given value for the control and outputs the value of the objective functional defined for it. It does this without explicitly depending on all the intermediary state variables, hence the name "reduced".
+#
+# To define the reduced functional, we provide the class with an objective (which is an overloaded UFL object) and the control. Both of these are essential for formulating the reduced functional.
+
+# +
+# Define the object for pyadjoint
+reduced_functional = ReducedFunctional(objective, control)
+# -
+
+# Pausing Annotation
+# ==================
+#
+# At this point, we have completed annotating the tape with the necessary information from running the forward simulation. To prevent further annotations during subsequent operations, we stop the annotation process. This ensures that no additional solves are unnecessarily recorded, keeping the tape focused only on the essential steps.
+#
+# We can then print the contents of the tape to verify what has been recorded.
```

Comment:
I was thinking plotting the tape instead of just printing the blocks. But it requires a lot of dependencies first, so I changed my mind. I might put back a print though.

Reply:
Did put back a print.